### DISCUSSION TITLE
emqx 5.8.3 setting acl in gui and having issues

### DISCUSSION THREAD

**Discussion by robertthaverly** [emqx/emqx#14385] [Q&A] [answered]
I have a THISTOPIC/#
for example I want users to be able to do as follows
username candoboth to be able to subscribe and publish to THISTOPIC/#  but no where else
username canpublish to be able to publish to THISTOPIC/#  but no where else nor any subscriptions
username cansubscribe to be able to subscribe to THISTOPIC/#  but no where else nor any other subscriptions/publish

I cant seem to get the combo correct, either they get no access or too much - 

%% Allow the MQTT client using username "dashboard" to subscribe to "$SYS/#" topics
{allow, {username, "dashboard"}, subscribe, ["$SYS/#"]}.

%% Allow users with IP address "127.0.0.1" to publish/subscribe to topics "$SYS/#", "#"
{allow, {ipaddr, "127.0.0.1"}, all, ["$SYS/#", "#"]}.

%% Deny all users from subscribing to sensitive topics: "$SYS/#", "#", and "+/#"
%% {deny, all, subscribe, ["$SYS/#", {eq, "#"}, {eq, "+/#"}]}.

%% candoboth User Rules
%% Allow "candoboth" to publish and subscribe only to "THISTOPIC/#"
{allow, {username, "candoboth"}, subscribe, ["THISTOPIC/#"]}.
{allow, {username, "candoboth"}, publish, ["THISTOPIC/#"]}.
%% Deny "candoboth" from accessing any other topics
%% {deny, {username, "candoboth"}, all, ["#"]}.

%% CanPublish User Rules
%% Allow "canpublish" to publish to "THISTOPIC/#" and deny all other actions
{allow, {username, "canpublish"}, publish, ["THISTOPIC/#"]}.
%% {deny, {username, "canpublish"}, all, ["#"]}.

%% CanSubscribe User Rules
%% Allow "cansubscribe" to subscribe to "THISTOPIC/#" and deny all other actions
{allow, {username, "cansubscribe"}, subscribe, ["THISTOPIC/#"]}.
%% {deny, {username, "cansubscribe"}, all, ["#"]}.

%% Deny all other publish and subscribe operations globally
%% Enforces a strict deny-all policy for security
%% {deny, all, publish, ["#"]}.
%% {deny, all, subscribe, ["#"]}.
%% {deny, all}.

**Comment by id** [comment]
@robertthaverly try to uncomment (remove leading `%%`) `{deny, all}.` at the bottom of the file.

**Comment by robertthaverly** [comment]
I had already tried that combo -- when I removed the comment %% {deny, all}. the first user candoboth
They can see $SYS  but appears they can only publish but not subscribe to THISTOPIC/#

I uncommented %% Deny all users from subscribing to sensitive topics: "$SYS/#", "#", and "+/#"
{deny, all, subscribe, ["$SYS/#", {eq, "#"}, {eq, "+/#"}]}.
That took care of the SYS I think  then tried to add a line with all instead of individual publish, and subscribe - this had no change
{allow, {username, "candoboth"}, all, ["THISTOPIC/#"]}.

highlighting the changes i made from above --  but in short the candoboth user can not subscribe and publish to the THISTOPIC/#
only publish

%% Deny all users from subscribing to sensitive topics: "$SYS/#", "#", and "+/#"
{deny, all, subscribe, ["$SYS/#", {eq, "#"}, {eq, "+/#"}]}.

%% candoboth User Rules
%% Allow "candoboth" to publish and subscribe only to "THISTOPIC/#"
{allow, {username, "candoboth"}, all, ["THISTOPIC/#"]}.
%% {allow, {username, "candoboth"}, subscribe, ["THISTOPIC/#"]}.
%% {allow, {username, "candoboth"}, publish, ["THISTOPIC/#"]}.
%% Deny "candoboth" from accessing any other topics
%% {deny, {username, "candoboth"}, all, ["#"]}.

{deny, all}.
=========================================================

2024-12-11T13:04:17.058351+00:00 [debug] tag: MQTT, clientid: mqtt-explorer-73aa8806, msg: mqtt_packet_sent, peername: 192.168.2.15:64433, username:candoboth, packet: SUBACK(Q0, R0, D0, PacketId=34960, ReasonCodes=[128])
2024-12-11T13:04:17.058685+00:00 [debug] clientid: mqtt-explorer-73aa8806, msg: insert_channel_info, peername: 192.168.2.15:64433, username:candoboth
2024-12-11T13:04:17.058766+00:00 [debug] tag: MQTT, clientid: mqtt-explorer-73aa8806, msg: mqtt_packet_sent, peername: 192.168.2.15:64433, username:candoboth, packet: SUBACK(Q0, R0, D0, PacketId=34961, ReasonCodes=[128])
2024-12-11T13:04:17.058928+00:00 [debug] clientid: mqtt-explorer-73aa8806, msg: insert_channel_info, peername: 192.168.2.15:64433, username:candoboth
2024-12-11T13:04:20.528658+00:00 [debug] tag: MQTT, clientid: mqtt-explorer-73aa8806, msg: raw_bin_received, peername: 192.168.2.15:64433, username:candoboth, size: 45, type: hex, bin: 312B001E696E7472696E736963732F74657374696E672D6465632D31312D323032346465632D31312D32303234
2024-12-11T13:04:20.528930+00:00 [debug] tag: MQTT, clientid: mqtt-explorer-73aa8806, msg: mqtt_packet_received, peername: 192.168.2.15:64433, username:candoboth, packet: PUBLISH(Q0, R1, D0, Topic=THISTOPIC/testing-dec-11-2024, PacketId=undefined, Payload(text)=dec-11-2024)
2024-12-11T13:04:20.529178+00:00 [debug] tag: AUTHZ, clientid: mqtt-explorer-73aa8806, msg: authorization_module_ignore, peername: 192.168.2.15:64433, username:candoboth, topic: THISTOPIC/testing-dec-11-2024, module: emqx_authz_client_info, action: PUBLISH(Q0,R1)
2024-12-11T13:04:20.529318+00:00 [debug] tag: AUTHZ, clientid: mqtt-explorer-73aa8806, msg: authorization_matched_allow, peername: 192.168.2.15:64433, username:candoboth, topic: THISTOPIC/testing-dec-11-2024, module: emqx_authz_file, action: PUBLISH(Q0,R1), authorize_type: file
2024-12-11T13:04:20.529418+00:00 [info] tag: AUTHZ, clientid: mqtt-explorer-73aa8806, msg: authorization_permission_allowed, peername: 192.168.2.15:64433, username:candoboth, topic: THISTOPIC/testing-dec-11-2024, action: PUBLISH(Q0,R1), source: file
2024-12-11T13:04:20.529506+00:00 [debug] tag: PUBLISH, clientid: mqtt-explorer-73aa8806, msg: publish_to, peername: 192.168.2.15:64433, username:candoboth, topic: THISTOPIC/testing-dec-11-2024, payload: dec-11-2024, payload_encode: text

**Comment by robertthaverly** [comment]
furthermore user cansubscribe - cannot subscribe and view
The only username that can subscribe is a super user only to any THISTOPIC/#

**Comment by id** [comment]
Just tested with emqx 5.8.3 and this acl file

```
{allow, {username, {re, "^dashboard$"}}, subscribe, ["$SYS/#"]}.

{allow, {ipaddr, "127.0.0.1"}, all, ["$SYS/#", "#"]}.

{deny, all, subscribe, ["$SYS/#", {eq, "#"}, {eq, "+/#"}]}.

{allow, {username, "candoboth"}, subscribe, ["THISTOPIC/#"]}.
{allow, {username, "candoboth"}, publish, ["THISTOPIC/#"]}.

{deny, all}.
```

```
$ mqttx sub -h 127.0.0.1 -t THISTOPIC/# -u candoboth
✔ Connected
✔ Subscribed to THISTOPIC/#
```

```
mqttx pub -h 127.0.0.1 -t THISTOPIC/1 -u candoboth
✔ Connected
✔ Message published
```

```
topic: THISTOPIC/1, qos: 0
Hello From MQTTX CLI
```

**Reply by id** [reply]
Correction: I needed to remove `{allow, {ipaddr, "127.0.0.1"}, all, ["$SYS/#", "#"]}.` line in order to block access for the other users. If you are running client on the same host as EMQX, this rule would effectively allow anyone to do anything.

**Reply by id** [reply, marked as answer]
corrected acl.conf
```
{allow, {username, {re, "^dashboard$"}}, subscribe, ["$SYS/#"]}.
{deny, all, subscribe, ["$SYS/#", {eq, "#"}, {eq, "+/#"}]}.

{allow, {username, "candoboth"}, subscribe, ["THISTOPIC/#"]}.
{allow, {username, "candoboth"}, publish, ["THISTOPIC/#"]}.

{allow, {username, "canpublish"}, publish, ["THISTOPIC/#"]}.
{allow, {username, "cansubscribe"}, subscribe, ["THISTOPIC/#"]}.

{deny, all}.
```

canpublish
```
$ mqttx sub -h 127.0.0.1 -t THISTOPIC/# -u canpublish
✔ Connected
✖ Subscription negated to "THISTOPIC/#" with code 135

$ mqttx pub -h 127.0.0.1 -t THISTOPIC/1 -u canpublish -q 1
✔ Connected
✔ Message published
```

cansubscribe
```
$ mqttx sub -h 127.0.0.1 -t THISTOPIC/# -u cansubscribe
✔ Connected
✔ Subscribed to THISTOPIC/#

$ mqttx pub -h 127.0.0.1 -t THISTOPIC/1 -u cansubscribe -q 1
✔ Connected
✖ Error: Publish error: Not authorized
```

**Comment by robertthaverly** [comment]
i updated as follows - but the results were the same as before on this mornings post so unchanged

candoboth can publish but not subscribe to this THISTOPIC/#
i think canpublish is set right THISTOPIC/#
cansubscribe - can not subscribe -- it cant publish as expected  THISTOPIC/#
again only one who seems to be able to subscribe are super users

```
%%-------------- Updated ACL Configuration -------------------

%% Allow the MQTT client using username "dashboard" to subscribe to "$SYS/#" topics
{allow, {username, "dashboard"}, subscribe, ["$SYS/#"]}.

%% Deny all users from subscribing to sensitive topics: "$SYS/#", "#", and "+/#"
{deny, all, subscribe, ["$SYS/#", {eq, "#"}, {eq, "+/#"}]}.

%% THISTOPIC User Rules
%% Allow "THISTOPIC" to publish and subscribe only to "THISTOPIC/#"
{allow, {username, "candoboth"}, subscribe, ["THISTOPIC/#"]}.
{allow, {username, "candoboth"}, publish, ["THISTOPIC/#"]}.

%% CanPublish User Rules
%% Allow "canpublish" to publish to "THISTOPIC/#" and deny all other actions
{allow, {username, "canpublish"}, publish, ["THISTOPIC/#"]}.

%% CanSubscribe User Rules
%% Allow "cansubscribe" to subscribe to "THISTOPIC/#" and deny all other actions
{allow, {username, "cansubscribe"}, subscribe, ["THISTOPIC/#"]}.

%% Deny all other publish and subscribe operations globally
%% Enforces a strict deny-all policy for security
{deny, all}.


```

**Reply by id** [reply]
I cannot reproduce your problem. What tool do you use to test this? And how exactly?

**Reply by id** [reply]
Try to reproduce your issue with https://mqttx.app/.
